torch/_dynamo/variables/functions.py

```diff
@@ -303,18 +303,40 @@ def python_type(self):
-        # rather than simple inlining. E.g, putting `call_method` op in FX graph for `forward` method
-        # since we ensure `forward` of allowed modules can be traced by AOT safely.
-        # Note this is not only for allowed modules, as user customized modules can extend from
-        # allowed modules but using parent's `forward` method, which is also covered by this branch.
+        # Here, we check if the method that we are calling belongs to the
+        # builtin (or allowed) nn modules.
+
+        # Context - We have historically avoided Dynamo tracing into the methods of
+        # builtin nn modules. Specifically for `forward` method of builtin
+        # modules, in the past we observed control-flow/hard-to-support python
+        # features, leading to unnecessary and frequent graph breaks. To avoid
+        # this, we just decided to directly put `call_module` for builtin
+        # nn.modules and `call_method` for the `forward` of builtin nn.modules.
+        # With AOT Autograd becoming an integral part of torch.compile stack,
+        # this became even a bigger moot point because AOT Autograd will skip
+        # over all those python features and give a clean aten-level graph.
+        #
+        # It is possible that these original motivations are no longer valid,
+        # and we could simplify the codepath now by removing the special case.
+        # However, it might be too late to do this because downstream users
+        # of Dynamo might have already written Fx Transformations assuming the
+        # presence of these builtin modules.
+
+        # In the following path, we check if the method belongs to the builtin
+        # nn module. If it does, we redirect it to NNModule special handling of
+        # methods. For example, TorchDynamo does not need to trace the
+        # `children` method, we can directly get the children by constructing
+        # the relevant VariableTracker. Again, it is possible to just let Dynamo
+        # trace the methods and construct the VariableTrackers, however there
+        # might be lots of graph breaks today. We can try to fix this in near
+        # future, because this does not impact the downstream users of Dynamo.
         if isinstance(self.obj, variables.NNModuleVariable):
-            module_attr = getattr(self.fn, "__module__", "")
-            if (
-                module_attr is not None
-                and module_attr.startswith("torch.nn.")
-                or self.is_constant
-            ):
+            # Extract the module this method belongs to. For example, a
+            # non-overriden `children` method of a user-defined nn module will
+            # still use the builtin nn module `children` impl. And therefore, we
+            # can redirect it to NNModuleVariable call_method.
+            module_where_fn_is_defined = inspect.getmodule(self.fn)
+            if module_where_fn_is_defined and is_allowed(module_where_fn_is_defined):
```

Comment:
I think this would break torchbench ```bert_pytorch``` model, where customized ```TokenEmbedding``` extend from allowed module but using allowed module's ```forward``` method. I think this is not a rare case.
https://github.com/codertimo/BERT-pytorch/blob/master/bert_pytorch/model/embedding/token.py#L4

Reply:
My assessment is not accurate, addressed by offline discussion.